src/groupeddataframe/fastaggregates.jl

```diff
@@ -157,24 +157,84 @@ function copyto_widen!(res::AbstractVector{T}, x::AbstractVector) where T
 end
 
 function groupreduce!(res::AbstractVector, f, op, condf, adjust, checkempty::Bool,
-                      incol::AbstractVector, gd::GroupedDataFrame)
+                      incol::AbstractVector, gd::GroupedDataFrame, nthreads::Integer)
     n = length(gd)
+    groups = gd.groups
     if adjust !== nothing || checkempty
         counts = zeros(Int, n)
     end
-    groups = gd.groups
-    @inbounds for i in eachindex(incol, groups)
-        gix = groups[i]
-        x = incol[i]
-        if gix > 0 && (condf === nothing || condf(x))
-            # this check should be optimized out if U is not Any
-            if eltype(res) === Any && !isassigned(res, gix)
-                res[gix] = f(x, gix)
-            else
-                res[gix] = op(res[gix], f(x, gix))
+    nt = min(nthreads, Threads.nthreads())
```

Comment:
Did you intentionally allow 0 or negative values of `nt` here? Maybe we should throw an error? Similarly you allow `true` and `false` passed as `nthreads`. Is this intentional?

Reply:
Yes we could be stricter, and that would prepare the possibility of switching to `0` as the default to choose automatically the optimal number of threads.

Not sure about `true` and `false`. We could restrict the argument to `Int` if that's simpler. (Otherwise since `true` would be equivalent to `1` it could be counter-intuitive.)

Reply:
We traditionally do `Union{Signed, Unsigned}` in other places to disallow `Bool` 😄. And I agree to disallow `<1` as then `0` can be used later.

Reply:
OK I've moved to `Int` for simplicity. I've also added a check.